### ISSUE TITLE
Fails to toggle

### ISSUE THREAD

**Issue by Fred-Vatin** [Joolee/userchrome-toggle#5]
Firefox | OS | Extension
------- | ---- | ---- 
102.0   | Win 10 | 1.2

This is how my userchrome starts. **It works as intended**.

```css
@import url("./CSS/userChrome-tab-tweaks.css") screen;

@namespace url("http://www.mozilla.org/keymaster/gatekeeper/there.is.only.xul");
@namespace html url("http://www.w3.org/1999/xhtml");

@-moz-document url(chrome://browser/content/browser.xul),
	url(chrome://browser/content/browser.xhtml),
	url(chrome://browser/content/places/bookmarksSidebar.xhtml),
	url(chrome://browser/content/webext-panels.xhtml),
	url(chrome://browser/content/places/places.xhtml)
{
	/* Hide tab bar */
	:root[titlepreface="᠎"] #titlebar {
		visibility: collapse;
	}
/* And so on… */
}
```

For testing purpose, I added the `:root[titlepreface="᠎"]` on the first rule only. I copied this magic CSS selector from the addon. VScode shows me the invisible UTF-8 characters. So that’s not the problem.
![image](https://user-images.githubusercontent.com/22337329/177679268-18211beb-b4d6-4bbe-86aa-8ad82155f989.png)
The CSS file is encoded as UTF-8 (no bom).

I applied changes in addon, saved my userChromme.css and restart FF. But when I click the UC button, it notifies it applies the rule but nothing happens. Is it supposed to comment my userChrome ? because it doesn’t.

Did I do something wrong ?

**Comment by pbaykalov** [comment]
when you write following in a CSS selector:
`selector1 selector2`
it means that anything `selector2` selects should be a descendant of at least something `selector1` selects. 

In your case I'm not sure about what the @-selectors are for. I have much simplier userChrome.css which only contains the rules on the top level and it works as intended. 

> Is it supposed to comment my userChrome ?

I do not understand your question. The extension only modifies the attribute of a window so that you can make your CSS depend on it.

**Comment by Fred-Vatin** [comment]
> when you write following in a CSS selector: `selector1 selector2` it means that anything `selector2` selects should be a descendant of at least something `selector1` selects.
> 
> In your case I'm not sure about what the @-selectors are for. I have much simplier userChrome.css which only contains the rules on the top level and it works as intended.
> 
> > Is it supposed to comment my userChrome ?
> 
> I do not understand your question. The extension only modifies the attribute of a window so that you can make your CSS depend on it.

The @-selectors are good practice in userChrome.css to avoid side effects. You only select what you change. The thing is clicking the toggle button doesn’t change anything in the userChrome.css.

**Comment by pbaykalov** [comment]
You can try using browser console to see what's wrong. 

>The thing is clicking the toggle button doesn’t change anything in the userChrome.css.

Should it?

**Comment by Joolee** [comment]
The extension doesn't change your userchrome.css itself so that shouldn't be the problem. In the past, an extension was able to modify the XUL structure itself so you could do very fancy and malicious things to the window. Since Mozilla opted to adopt the WebExtension standard, that wasn't possible any more. The only thing I can modify with the extension is a sort of 'prefix' for the window title. That was designed to be able to develop extensions that let you label your Firefox Windows. Luckily, this change happens on the top-level of the XUL structure so you can use fancy CSS rules to apply only to windows having a certain title prefix.

To get to the point; does your CSS code actually remove the tab-bar when you remove only the `[titlepreface="᠎"]` part? I suspect not. Otherwise, I can't really think of any other problems you might be running in to.

**Comment by Fred-Vatin** [comment]
> To get to the point; does your CSS code actually remove the tab-bar when you remove only the `[titlepreface="᠎"]` part? 

Of course it works. Pretty easy to test by yourself through `ctrl+shift+alt+i` > style editor > edit userChrome.css and add

```css
#titlebar {
    visibility: collapse !important;
}
```

Then `ctrl+S` to save the css and return to the firefox UI.

I wrote my bug report so anyone can reproduce it.

**Comment by pbaykalov** [comment]
> I wrote my bug report so anyone can reproduce it.

Please answer my question.

**Comment by Fred-Vatin** [comment]
> > I wrote my bug report so anyone can reproduce it.
> 
> Please answer my question.

I wrote my answer : “**Of course it works**. Pretty easy to test by yourself through ctrl+shift+alt+i > style editor > edit userChrome.css and add.”

I don’t know how I can be clearer.

**Comment by pbaykalov** [comment]
>The thing is clicking the toggle button doesn’t change anything in the userChrome.css.

Should it?

**Comment by Fred-Vatin** [comment]
> > The thing is clicking the toggle button doesn’t change anything in the userChrome.css.
> 
> Should it?

Please can you stop spamming this issue asking the same questions again and again?

**Comment by pbaykalov** [comment]
> > To get to the point; does your CSS code actually remove the tab-bar when you remove only the `[titlepreface="᠎"]` part?
> 
> Of course it works. Pretty easy to test by yourself through `ctrl+shift+alt+i` > style editor > edit userChrome.css and add
> 
> ```css
> #titlebar {
>     visibility: collapse !important;
> }
> ```
> 
> Then `ctrl+S` to save the css and return to the firefox UI.
> 
> I wrote my bug report so anyone can reproduce it.

Tried to reproduce. Style toggling works as intended -- the bar hides and appears.

**Comment by pbaykalov** [comment]
![изображение](https://user-images.githubusercontent.com/54845956/195279853-6de3d07c-be4e-4714-88f7-c38de4621b6e.png)

![изображение](https://user-images.githubusercontent.com/54845956/195279891-94d9db59-30db-44d6-b994-8838066ec58c.png)

**Comment by pbaykalov** [comment]
But toggling does not work when rule is inside the @-block.

![изображение](https://user-images.githubusercontent.com/54845956/195281516-5a128f36-292c-4d88-879a-7e8e33f0291e.png)

**Comment by Joolee** [comment]
I'm not near a computer to test or modify anything this month so thank you for testing it for me! If there really is a problem when using @moz-document selectors, it's probably good to mention that in the addon description or on the settings page.

**Comment by pbaykalov** [comment]
Additionally:
https://developer.mozilla.org/en-US/docs/Web/CSS/@document - deprecated

~~Browser console starts spewing out error if I use this userchrome.~~

**Comment by Joolee** [comment]
That looks like a Firefox bug to me. Might be worth checking if there is already a bug report or open one. Though I doubt someone will fix it as userchrome is being deprecated as far as I know.

**Comment by pbaykalov** [comment]
Actually, the errors are there regardless of what userChrome I use.
Regardless, it also looks as a bug at least with this @-rule to me.

**Comment by Fred-Vatin** [comment]
> But toggling does not work when rule is inside the @-block.

Interesting. I will see if I can make my userChrome work putting this rule outside the @-block.

**Comment by Fred-Vatin** [comment]
This works
```css
@import url("./CSS/userChrome-tab-tweaks.css") screen;

/* Hide tab bar */
:root[titlepreface*="᠎"] #titlebar {
    visibility: collapse !important;
}

@namespace url("http://www.mozilla.org/keymaster/gatekeeper/there.is.only.xul");
@namespace html url("http://www.w3.org/1999/xhtml");
```

This does NOT work although it’s not inside a @-block but after the `@namespace`
```css
@import url("./CSS/userChrome-tab-tweaks.css") screen;

@namespace url("http://www.mozilla.org/keymaster/gatekeeper/there.is.only.xul");
@namespace html url("http://www.w3.org/1999/xhtml");

/* Hide tab bar */
:root[titlepreface*="᠎"] #titlebar {
    visibility: collapse !important;
}
```

But we would want to omit `@namespace` as discussed [here](https://www.userchrome.org/adding-style-recipes-userchrome-css.html#:~:text=ignored.%20For%20example%3A-,recommendation,-It%20probably%20is).

`@-moz-document` block is still required to style some elements like other windows or extensions though.

In a nutshell, for the extension works you need this template…

```css
/* Here if you need to import some external css */
@import url("./CSS/userChrome-some-addons.css") screen;

/* Here all the rules can be toggled with Userchrome Toggle */
:root[titlepreface*="᠎"] #titlebar {
    visibility: collapse !important;
}

/* Here all specific rules for user interface, extensions UI… */
@-moz-document url(chrome://browser/content/browser.xul),
	       url(chrome://browser/content/browser.xhtml),
	       url(chrome://browser/content/places/bookmarksSidebar.xhtml),
	       url(chrome://browser/content/webext-panels.xhtml),
	       url(chrome://browser/content/places/places.xhtml)
{
    /* Your style rules for the user interface go here */
    …
}
```

**Comment by krystian3w** [comment]
Remove namespace:

https://www.userchrome.org/adding-style-recipes-userchrome-css.html

>But is that really true? What does it do?
>
> **tl;dr** - You do not need to use it, and if you do, there might be some unexpected glitches.

> It probably is possible to write style rules to apply to HTML elements when only the XUL namespace is specified, but it's simpler not to create this situation in the first place and omit all mention of namespaces.
